interpreter/runtimecost.go

```diff
@@ -0,0 +1,184 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package interpreter provides functions to evaluate parsed expressions with
+// the option to augment the evaluation with inputs and functions supplied at
+// evaluation time.
+
+package interpreter
+
+import (
+	"github.com/google/cel-go/checker"
+	"github.com/google/cel-go/common/overloads"
+	"github.com/google/cel-go/common/types/ref"
+	"math"
+	"reflect"
+)
+
+// ActualCostEstimator is the interface to calculate cost for wired in function
+type ActualCostEstimator interface {
+	CallCost(overloadId string, args []ref.Val) *uint64
+}
+
+// CostTracker represents the information needed for tacking runtime cost
+type CostTracker struct {
+	estimator         *checker.CostEstimator
+	CallCostEstimator ActualCostEstimator
+	state             EvalState
+	cost              uint64
+}
+
+// actualSize returns the size of value
+func (c CostTracker) actualSize(i Interpretable, ctx Activation) uint64 {
+	if v, ok := c.state.Value(i.ID()); ok {
+		reflectV := reflect.ValueOf(v.Value())
+		switch reflectV.Kind() {
+		// Note that the CEL bytes type is implemented with Go byte slices, therefore also supported by the following
+		// code.
+		case reflect.String, reflect.Array, reflect.Slice, reflect.Map:
+			return uint64(reflectV.Len())
+		}
+	}
+	return 0
+}
+
+// ActualCost returns the runtime cost
+func (c CostTracker) ActualCost() uint64 {
+	return c.cost
+}
+
+type evalCostTrackerConst struct {
+	InterpretableConst
+	tracker *CostTracker
+}
+
+func (e *evalCostTrackerConst) Eval(ctx Activation) ref.Val {
+	val := e.InterpretableConst.Eval(ctx)
+	return val
+}
+
+type evalCostTrackerAttribute struct {
+	InterpretableAttribute
+	tracker *CostTracker
+}
+
+func (e *evalCostTrackerAttribute) Eval(ctx Activation) ref.Val {
+	_, maxCost := estimateCost(e.InterpretableAttribute)
+	if maxCost > 0 {
+		e.tracker.cost += uint64(maxCost)
+	}
+	val := e.InterpretableAttribute.Eval(ctx)
+	return val
+}
+
+type evalCostTrackerCall struct {
+	InterpretableCall
+	tracker *CostTracker
+}
+
+func (e *evalCostTrackerCall) Eval(ctx Activation) ref.Val {
+	val := e.InterpretableCall.Eval(ctx)
+	args := e.InterpretableCall.Args()
+	if e.tracker.CallCostEstimator != nil {
+		var argValues []ref.Val
+		for _, arg := range args {
+			if v, ok := e.tracker.state.Value(arg.ID()); ok {
+				argValues = append(argValues, v)
+			} else {
+				// FIXME: abort.
+			}
+		}
+		e.tracker.cost += *e.tracker.CallCostEstimator.CallCost(e.OverloadID(), argValues)
```

Comment:
I was expecting CallCostEstimator to take precedence over the built in cost code, but to fall back to it if it returns nil. E.g.

```
cost := e.tracker.CallCostEstimator.CallCost(...)
if cost != nil {
  e.tracker.cost += *cost
  return val
}
// fallback to switch e.OverloadID()...
```